source/AI.cpp

```diff
@@ -2594,6 +2639,113 @@ bool AI::DoCloak(Ship &ship, Command &command)
 
 
 
+bool AI::DoRoving(Ship &ship, Command &command, unsigned int rovingShipCount)
+{	
+	const Point target = ship.GetTargetPosition();
+	const auto v = ship.MaxVelocity();
+	if(!target || MoveTo(ship, command, target, Point(), v, v))
+	{
+		Point newTarget = Angle::Random().Unit() * Random::Real() * MAX_DISTANCE_FROM_CENTER;
+		ship.SetTargetPosition(newTarget);
+	}
+	
+	const System *system = ship.GetSystem();
+	if(rovingShipCount > 9 && system)
+		return DoBored(ship, command, system);
+	
+	return false;
+}
+
+
+
+bool AI::DoEvasive(Ship &ship, Command &command)
+{
+	// Did I have to move my targetPosition?
+	bool movedTarget = false;
+	
+	// First, figure out which ships are not of the same government. Ignore disabled/destroyed ships.
+	vector<shared_ptr<Ship>> enemies;
+	for(auto otherShip : GetShipsList(ship, false))
+		if(otherShip->GetGovernment() != ship.GetGovernment())
+			enemies.push_back(make_shared<Ship>(*otherShip));
+	
+	// Loop through enemies and figure out their range and how to evade them.
+	vector<Point> paths;
+	double maxRange = 0.;
+	const Point pos = ship.Position();
+	for(const auto &enemy : enemies)
+	{
+		// Ignore disabled/destroyed ships.
+		if(enemy->IsDisabled() || enemy->IsDestroyed())
+			continue;
+		
+		// Find out their weapons range.
+		for(const auto &weapon : enemy->Weapons())
+			maxRange = max(maxRange, weapon.GetOutfit()->Range());
+		maxRange *= 2;
```

Comment:
why do we multiply it by 2? I think multiplying it by 1.5 would be reasonable, or maybe we should instead just add what we consider to be a safe distance to it; like idk, say 500

Reply:
I played around with the numbers and found that if the range is multiplied by 2, they have enough time to turn around and run. I tried 1.5 and 1.2, both were too slow.

Reply:
Could you add a comment explaining it?
What I was suggesting was defining a safe distance (say 500) and adding it to the range, because rn if a ship has huge range your ship will go too far, and if it uses firethrower you'll be too close.

Reply:
I tried adding a few numbers to the range like you suggested, so far adding 1,000 seems to give them enough time to get out of the way. I'll push the changes once I resolve the other suggestions you made.